spec/index.bs

```diff
@@ -3522,8 +3526,125 @@ The following validation rules apply:
+
+Debug groups in a {{GPUCommandEncoder}} or {{GPUProgrammablePassEncoder}}
+must be well nested.
+
+### <dfn method for=GPUCommandEncoder>pushDebugGroup(groupLabel)</dfn> ### {#GPUCommandEncoder-pushDebugGroup}
+
+<div algorithm="GPUCommandEncoder.pushDebugGroup">
+    <strong>|this|:</strong> of type {{GPUCommandEncoder}}.
+
+    **Arguments:**
+      - {{DOMString}} |groupLabel|
+
+    **Returns:** void
+
+    Marks the beginning of a labeled group of commands for the {{GPUCommandEncoder}}.
+
+    |groupLabel| defines the label for the command group.
+
+    On the [=Device timeline=], the following steps occur:
+
+        - If the user agent can fulfill the request and
+            the [$GPUCommandEncoder.pushDebugGroup/Valid Usage$] rules are met:
+
+            - push |groupLabel| onto then end of |this|.{{GPUCommandEncoder/[[debug_group_stack]]}}.
+
+    <div class=validusage dfn-for=GPUCommandEncoder.pushDebugGroup>
+        <dfn abstract-op>Valid Usage</dfn>
+
+        For |this|:
+        - |this|.{{GPUCommandEncoder/[[state]]}} must be {{encoder state/open}}.
+    </div>
+
+</div>
+
+### <dfn method for=GPUCommandEncoder>popDebugGroup()</dfn> ### {#GPUCommandEncoder-popDebugGroup}
+
+<div algorithm="GPUCommandEncoder.popDebugGroup">
+    <strong>|this|:</strong> of type {{GPUCommandEncoder}}.
+
+    **Returns:** void
+
+    Marks the end of a labeled group of commands for the {{GPUCommandEncoder}}.
+
+    On the [=Device timeline=], the following steps occur:
+
+        - If the user agent can fulfill the request and
+            the [$GPUCommandEncoder.popDebugGroup/Valid Usage$] rules are met:
+
+            - pop an entry off the end of |this|.{{GPUCommandEncoder/[[debug_group_stack]]}}.
+
+    <div class=validusage dfn-for=GPUCommandEncoder.popDebugGroup>
+        <dfn abstract-op>Valid Usage</dfn>
+
+        For |this|:
+        - |this|.{{GPUCommandEncoder/[[state]]}} must be {{encoder state/open}}.
+        - |this|.{{GPUCommandEncoder/[[debug_group_stack]]}}.length must be greater than 0.
+    </div>
+
+</div>
+
+### <dfn method for=GPUCommandEncoder>insertDebugMarker(markerLabel)</dfn> ### {#GPUCommandEncoder-insertDebugMarker}
+
+<div algorithm="GPUCommandEncoder.insertDebugMarker">
+    <strong>|this|:</strong> of type {{GPUCommandEncoder}}.
+
+    **Arguments:**
+        - {{DOMString}} |markerLabel|
+
+    **Returns:** void
+
+    Inserts a single debug marker label into the {{GPUCommandEncoder}}'s commands sequence .
+
+    |markerLabel| defines the label to insert.
+
+    <div class=validusage dfn-for=GPUCommandEncoder.insertDebugMarker>
+        <dfn abstract-op>Valid Usage</dfn>
+
+        For |this|:
+        - |this|.{{GPUCommandEncoder/[[state]]}} must be {{encoder state/open}}.
+    </div>
+
+</div>
+
+## Finalization ## {#command-encoder-finalization}
+
+Once the command encoder has finished submitting any commands a {{GPUCommandBuffer}} can be created
```

Comment:
let's not use the verb "submitting" here in order to avoid confusion. Perhaps, we could say "recording"?

Reply:
Also, it reads unclear to me. How does the user know if "command encoder has finished xxx"?
We should rewrite it from the user's point of view, e.g. "a command encoder can be closed and trasformed into a command buffer by calling XXx."

Reply:
Took a stab at this, let me know if the rewritten version reads better to you.